incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -38,25 +59,306 @@ type ObjectReconcilerNew struct {
+		if !errors.IsNotFound(err) {
+			log.Info("Couldn't read")
+			return resp, err
+		}
+		found = false
+	}
+	act, so := r.syncWithForest(ctx, log, found, inst)
+
+	return resp, r.operate(ctx, log, act, inst, so)
+}
+
+// syncWithForest syncs the object instance with the in-memory forest. This is the only place we should
+// lock the forest in each Reconcile, so this fn needs to return everything relevant for the rest of the
+// Reconcile. This fn shouldn't contact the apiserver since that's a slow operation and everything will
+// block on the lock being held. It returns the action to take on the object (delete, overwrite or do
+// nothing) and a source object if the action is to overwrite it.
+func (r *ObjectReconcilerNew) syncWithForest(ctx context.Context, log logr.Logger, found bool, inst *unstructured.Unstructured) (action, *unstructured.Unstructured) {
+	r.Forest.Lock()
+	defer r.Forest.Unlock()
+
+	if found {
+		// If it's about to be deleted, do nothing, just wait for it to be actually deleted.
+		if !inst.GetDeletionTimestamp().IsZero() {
+			log.Info("Will soon be deleted")
+			return unchanging, nil
+		}
+		return r.syncObject(ctx, log, inst)
+	}
+
+	// If this is a source that was deleted, update the forest and enqueue all descendants.
+	if r.isSource(inst) {
+		log.Info("Deleting the source copy in the forest")
+		r.removeDeletedSource(inst)
+		log.Info("Enqueuing all descendants of the deleting source")
+		r.enqueueDescendants(ctx, log, inst)
+		return unchanging, nil
+	}
+
+	// This is a non-existent yet propagated object. Set "api.LabelInheritedFrom" label.
+	// The correct value will be set in the "overwrite" function.
+	metadata.SetLabel(inst, api.LabelInheritedFrom, "sns placeholder")
+	log.Info("Reconciling non-existent object")
+	return r.syncObject(ctx, log, inst)
+}
+
+// syncObject handles a source object and a propagated object differently. If a source object changes,
+// all descendant copies will be enqueued. If a propagated object is obsolete, it will be deleted.
+// Otherwise, it will be overwritten by the source if they are different.
+func (r *ObjectReconcilerNew) syncObject(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (action, *unstructured.Unstructured) {
+	// If for some reason this has been called on an object that isn't namespaced, let's generate some
+	// logspam!
+	if inst.GetNamespace() == "" {
+		for i := 0; i < 100; i++ {
+			log.Info("Non-namespaced object!!!")
+		}
+		return unchanging, nil
+	}
+
+	// This object is the source.
+	if !isPropagated(inst) {
+		log.Info("This object is the source", "namespace", inst.GetNamespace())
+		r.syncSource(ctx, log, inst)
+		// Early exit.
+		return unchanging, nil
+	}
+
+	// This object is a propagated copy.
+	log.Info("This object is a propagated copy", "namespace", inst.GetNamespace())
+	return r.syncPropagated(ctx, log, inst)
+}
+
+// syncPropagated will delete obsolete copy or overwrite itself with the source.
```

Comment:
It doesn't actually delete/overwrite - maybe "determine whether to delete or overwrite"?